certregistry/certregistry.go

```diff
@@ -0,0 +1,98 @@
+package certregistry
+
+import (
+	"crypto/tls"
+	"errors"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	errCertNotFound = errors.New("certificate not found")
+	defaultHost     = "ingress.local"
+)
+
+type tlsCertificate struct {
+	hosts []string
+	cert  *tls.Certificate
+}
+
+// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// We ensure ensure syncronized access to them and hold a default certificate. 
+type CertRegistry struct {
+	lookup         map[string]*tlsCertificate
+	mx             *sync.Mutex
+	defaultTLSCert *tls.Certificate
+}
+
+// NewCertRegistry initializes the certificate registry with an empty map 
+// and a generated default certificate. 
+func NewCertRegistry() *CertRegistry {
+	l := make(map[string]*tlsCertificate)
+
+	return &CertRegistry{
+		lookup:         l,
+		mx:             &sync.Mutex{},
+		defaultTLSCert: getFakeHostTLSCert(defaultHost),
+	}
+}
+
+func (r *CertRegistry) getCertByKey(key string) (*tlsCertificate, error) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	cert, ok := r.lookup[key]
+	if !ok || cert == nil {
+		log.Debugf("certificate not found in registry - %s", key)
+		return nil, errCertNotFound
+	}
+	
+	return cert, nil
+}
+
+func (r *CertRegistry) addCertToRegistry(key string, cert *tlsCertificate) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	r.lookup[key] = cert
+}
+
+// SyncCert takes a TLS certificate and list of hosts and saves them to the registry with the 
+// provided key. If the object already exists it will be updated or added otherwise. Returns 
+// true if key was changed.
+func (r *CertRegistry) SyncCert(key string, hosts []string, crt *tls.Certificate) bool {
+	cert := &tlsCertificate{
+		hosts: hosts, 
+		cert: crt,
+	}	
+
+	curr, err := r.getCertByKey(key)
+	if err == nil {
```

Comment:
If we get `err != nil`, we will just go to "adding cert" case

Reply:
Right, `if err != nil` it means the cert was not found with that key so we'll add it to the registry. I could use an `else` statement instead but it seemed redundant.

Reply:
Improved the logic here so it's more clear what happens when we return `err` or not.

Reply:
I am not sure if I understand the logic here. :)
Why do we add a certificate that we don't found into our registry?
Do we need to refresh it for some reason?

Reply:
If we ignore it, we should not return error but bool to highlight "update required, as you know from map lookup for example:
```
if curr, ok := r.getCertByKey(key); ok {
    if !equalCert(curr, cert) {
	log.Debugf("updating certificate in registry - %s", key)
	r.addCertToRegistry(key, cert)
        return true
    } else {
        return false
    }
}

log.Debugf("adding new certificate to registry - %s", key)
r.addCertToRegistry(key, cert)
```

Reply:
I see what you mean. Updated to return `bool` instead of `error`.